File: gemstone/src/models/balance.rs
use num_bigint::BigUint;
use primitives::{AssetBalance, AssetId, Balance, asset_balance::BalanceMetadata};

pub type GemAssetBalance = AssetBalance;
pub type GemBalance = Balance;
pub type GemBalanceMetadata = BalanceMetadata;

#[uniffi::remote(Record)]
pub struct GemAssetBalance {
    pub asset_id: AssetId,
    pub balance: GemBalance,
    pub is_active: bool,
}

#[uniffi::remote(Record)]
pub struct GemBalance {
    pub available: BigUint,
    pub frozen: BigUint,
    pub locked: BigUint,
    pub staked: BigUint,
    pub pending: BigUint,
    pub rewards: BigUint,
    pub reserved: BigUint,
    pub withdrawable: BigUint,
    pub metadata: Option<GemBalanceMetadata>,
}

#[uniffi::remote(Record)]
pub struct GemBalanceMetadata {
    pub energy_available: u64,
    pub energy_total: u64,
    pub bandwidth_available: u64,
    pub bandwidth_total: u64,
}

File: Packages/Primitives/Sources/Balance.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import BigInt

public struct Balance: Codable, Equatable, Hashable, Sendable {
    public var available: BigInt
    public var frozen: BigInt
    public var locked: BigInt
    public var staked: BigInt
    public var pending: BigInt
    public var pendingUnconfirmed: BigInt
    public var rewards: BigInt
    public var reserved: BigInt
    public var withdrawable: BigInt
    public var earn: BigInt
    public var metadata: BalanceMetadata?

    public init(
        available: BigInt = .zero,
        frozen: BigInt = .zero,
        locked: BigInt = .zero,
        staked: BigInt = .zero,
        pending: BigInt = .zero,
        pendingUnconfirmed: BigInt = .zero,
        rewards: BigInt = .zero,
        reserved: BigInt = .zero,
        withdrawable: BigInt = .zero,
        earn: BigInt = .zero,
        metadata: BalanceMetadata? = .none
    ) {
        self.available = available
        self.frozen = frozen
        self.locked = locked
        self.staked = staked
        self.pending = pending
        self.pendingUnconfirmed = pendingUnconfirmed
        self.rewards = rewards
        self.reserved = reserved
        self.withdrawable = withdrawable
        self.earn = earn
        self.metadata = metadata
    }
}
